common/src/main/java/org/terracotta/angela/common/distribution/Distribution.java

```diff
@@ -92,29 +85,59 @@ public LicenseType getLicenseType() {
     return licenseType;
   }
 
-  public DistributionController createDistributionController() {
-    if (version.getMajor() >= 11
-        || (version.getMajor() == 10 && version.getMinor() >= 7) //tc-ee 10.7
-        || (version.getMajor() == 5 && version.getMinor() >= 7) //tc-platform 5.7 and above
-        || (version.getMajor() == 3 && version.getMinor() >= 9) //ehcache 3.9 and above
-    ) {
-      if (runtimeOptions.contains(RuntimeOption.INLINE_SERVERS)) {
-        return new Distribution107InlineController(this);
-      } else {
-        return new Distribution107Controller(this);
-      }
+  public Class<? extends DistributionController> getDistributionControllerType() {
+    // manually set ?
+    if (distributionControllerType != null) {
+      return distributionControllerType;
+    }
+
+    // 11.x, 12.x and above => TCDB
+    if (version.getMajor() >= 11) {
+      return runtimeOptions.contains(RuntimeOption.INLINE_SERVERS) ? Distribution107InlineController.class : Distribution107Controller.class;
     }
 
-    if (version.getMajor() == 10 || version.getMajor() == 3) {
-      return new Distribution102Controller(this);
-    } else {
-      if (version.getMinor() >= 3) {
-        return new Distribution43Controller(this);
-      }
+    // 10.7 and above => TCDB
```

Comment:
Shouldn't there be a cleaner way to determine the desired controller? What happens if if we move to Ehcache 4? It would seem this would require and {productName, version} pair.

Reply:
I agree with you, that is why I introduced the override `.withDistributionController ()`.

The checks are roughly what they were before but easier to read now